src/Markup/Avalonia.Markup.Xaml/XamlIl/CompilerExtensions/Transformers/AvaloniaXamlIlDataContextTypeTransformer.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Avalonia.Markup.Parsers;
+using Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Transformers;
+using Avalonia.Utilities;
+using XamlIl;
+using XamlIl.Ast;
+using XamlIl.Transform;
+using XamlIl.TypeSystem;
+
+namespace Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Transformers
+{
+    class AvaloniaXamlIlDataContextTypeTransformer : IXamlIlAstTransformer
+    {
+        private const string AvaloniaNs = "https://github.com/avaloniaui";
+        public IXamlIlAstNode Transform(XamlIlAstTransformationContext context, IXamlIlAstNode node)
+        {
+            if (node is XamlIlAstObjectNode on)
+            {
+                AvaloniaXamlIlDataContextTypeMetadataNode calculatedDataContextTypeNode = null;
+                foreach (var child in on.Children)
+                {
+                    if (child is XamlIlAstXmlDirective directive)
+                    {
+                        if (directive.Namespace == XamlNamespaces.Xaml2006
+                            && directive.Name == "DataContextType"
+                            && directive.Values.Count == 1
+                            && directive.Values[0] is XamlIlTypeExtensionNode dataContextType)
+                        {
+                            on.Children.Remove(child);
+                            return new AvaloniaXamlIlDataContextTypeMetadataNode(on, dataContextType.Value.GetClrType());
+                        }
+                    }
+                    else if (child is XamlIlAstXamlPropertyValueNode pv
+                        && pv.Property is XamlIlAstNamePropertyReference pref
+                        && pref.Name == "DataContext"
+                        && pref.DeclaringType is XamlIlAstXmlTypeReference tref
+                        && tref.Name == "StyledElement"
+                        && tref.XmlNamespace == AvaloniaNs)
+                    {
+                        var bindingType = context.GetAvaloniaTypes().IBinding;
+                        if (!pv.Values[0].Type.GetClrType().GetAllInterfaces().Contains(bindingType))
+                        {
+                            calculatedDataContextTypeNode = new AvaloniaXamlIlDataContextTypeMetadataNode(on, pv.Values[0].Type.GetClrType());
+                        }
+                        else if(pv.Values[0].Type.GetClrType().Equals(context.GetAvaloniaTypes().CompiledBindingExtension)
```

Comment:
Even if we pre-parse the compiled bindings before evaluating, that won’t allow us to infer DataContext’s of referenced elements without a multi-pass transformer. Users could set it up that control A’s data context depends on B’s data context which depends on C’s data context in such a way that our current DFS transformer algorithm (or a BFS) would be unable to resolve in a single pass.

I’d rather leave that to a second PR since this one is already very long.

Reply:
You can save arbitrary data in XamlIlTransformationContext via SetItem/GetItem. So one the first transformer would populate some data structure with the full list of bindings, then another one can manipulate the data as it sees fit. Another type of transformers that accepts the whole document instead of a tree node might better suit that task.


It's fine to leave that to a separate PR.